src/lib/optimizer/strategy/between_composition_rule.cpp

```diff
@@ -330,53 +329,81 @@ void BetweenCompositionRule::_replace_predicates(const std::vector<std::shared_p
     }
   }
 
-  // Append between nodes to predicate nodes to get the complete chain of all necessary LQP nodes
+  // Append between nodes to PredicateNodes to get the complete chain of all necessary LQP nodes
   predicate_nodes.insert(predicate_nodes.cend(), between_nodes.cbegin(), between_nodes.cend());
 
-  // Insert predicate nodes to LQP
-  // Connect last predicate to chain input
-  predicate_nodes.back()->set_left_input(input);
+  // Insert PredicateNodes into the LQP
+  //  Connect first PredicateNode to chain input
+  predicate_nodes.front()->set_left_input(predicate_chain_input_node);
 
-  // Connect predicates
-  for (size_t predicate_index = 0; predicate_index < predicate_nodes.size() - 1; predicate_index++) {
-    predicate_nodes[predicate_index]->set_left_input(predicate_nodes[predicate_index + 1]);
+  //  Connect PredicateNodes among each other
+  for (auto predicate_index = size_t{1}; predicate_index < predicate_nodes.size(); ++predicate_index) {
+    predicate_nodes[predicate_index]->set_left_input(predicate_nodes[predicate_index - 1]);
   }
 
-  // Connect first predicates to chain output
-  for (size_t output_index = 0; output_index < outputs.size(); ++output_index) {
-    outputs[output_index]->set_input(input_sides[output_index], predicate_nodes.front());
+  //  Connect last PredicateNode with all output nodes of the chain
+  for (size_t output_index = 0; output_index < predicate_chain_output_nodes.size(); ++output_index) {
+    predicate_chain_output_nodes[output_index]->set_input(predicate_chain_output_input_sides[output_index],
+                                                          predicate_nodes.back());
   }
 }
 
 void BetweenCompositionRule::_apply_to_plan_without_subqueries(const std::shared_ptr<AbstractLQPNode>& lqp_root) const {
-  if (lqp_root->type == LQPNodeType::Predicate) {
-    std::vector<std::shared_ptr<PredicateNode>> predicate_nodes;
-
-    // Gather adjacent PredicateNodes
-    auto current_node = lqp_root;
-    while (current_node->type == LQPNodeType::Predicate) {
-      // Once a node has multiple outputs, we're not talking about a predicate chain anymore
+  std::unordered_set<std::shared_ptr<PredicateNode>> visited_predicate_nodes;
+  std::unordered_set<std::shared_ptr<AbstractLQPNode>> visited_nodes;
+  std::vector<std::vector<std::shared_ptr<PredicateNode>>> grouped_predicate_nodes;
```

Comment:
Add a comment: Grouped by what?

Also, I am starting to lose track of what (a) adjacent, (b) chained, and (c) grouped mean.